### PR TITLE
Update humidity ratio calculation in PsyTwbFnTdbWPb() function

### LINKED ISSUE #7187: PsyTwbFnTdbWPb: humidity ratio calculation does not include case for TWetBulb < 0
This may not be an issue per se but I just wanted to report it given that I have noticed differences between the results of wet-bulb temperature from dry-bulb temperature, humidity ratio, and pressure calculated by EnergyPlus and [PsychroLib](https://github.com/psychrometrics/psychrolib#readme) when running tests for a separate model I am working with. The following snippet:

https://github.com/NREL/EnergyPlus/blob/d830726d9ea59782c5f19a6ceab71219106e7713/src/EnergyPlus/Psychrometrics.cc#L661-L663

shows that the humidity ratio is calculated for cases where the temperature is below and above zero in the same way. The ASHRAE Handbook however also gives a formulae with different constants for cases where the wet-bulb temperature is below zero as also done in PsychroLib -- see [this](https://github.com/psychrometrics/psychrolib/blob/517686dc7f5c6caef5a79c653dfb2ed563dec7bf/src/fortran/psychrolib.f90#L573-L578). This difference seem to amount to a max wet-bulb temperature difference of approximately 0.7 degree C close to zero, exponentially decreasing to approximately 0.1 at about -30 degree C dry bulb temperature. If you think that this may be worth addressing, I am happy to send you a PR for the additional case when the temperature is below zero.

### PR DESCRIPTION
**Pull request by dareumnam**
Pull request overview
---------------------
 - Fixes #7187
 - Psychrometrics.cc
   - In PsyTwbFnTdbWPb() function, humidity ratio calculation must be separated for cases where the wet-bulb temperature is below zero or not.
   - The current equation was according to **ASHRAE Handbook Fundamentals (1997)**, SI unit edition (ch.6, eqn (35)), in the picture below
image
   - It was updated according to **ASHRAE Handbook Fundamentals (2017)**, SI unit edition (ch.1, eqn (35) and (37)), in the picture below
image

- Psychrometrics.unit.cc
   - Add a unit test case: use PsyTwbFnTdbWPb() function when expected result of wet-bulb temperature is below zero
   - According to Psychrometric chart, T_wb is -2.2C when 
     - T_db = 1C
     - W = 0.002 kg.w/kg.da
     - P = 101325 Pa  
   - With the previous equation, PsyTwbFnTdbWPb(1C, 0.002, 101325 Pa ) = -1.9891, which is wrong according to the psychrometric chart.
   - With the update equation, PsyTwbFnTdbWPb(1C, 0.002, 101325 Pa ) = -2.1992, which is right accoridng to the psychrometric chart.
image
image

- Plot the function to check if the graph is smooth where WBT is around 0 
  - According to the psychrometric chart, negative wet-bulb temperatures exist where the dry-bulb temperatures are between 0°C to 10.5°C (inside of the yellow triangle in the picture below).
image
  - So I called PsyTwbFnTdbWPb(Tdb, W, Pb) function with arguments of
    - Dry-bulb Temp. of 0°C to 10°C (fixed value in each call)
    - Humidity ratio that is equivalent to Relative Humidity of 0% to 100% (X-axis)
    - Atmospheric pressure, 101325 Pa (fixed value in each call)
  - And plotted Wet-bulb temperature (Y-axis).
  - The plotting results are as below. The result showed some gaps where the wet-blub temperature is 0°C, particularly where the dry-bulb temperature is getting farther from 0°C. The equations might not contain the special trend on the Psychrometric chart where the wet-bulb temp = 0.
image
    - I am pretty sure that the equation change itself of this PR is correct based on the ASHRAE handbook update. My best guess for these gaps is: 
According to the Psychrometric Chart in the ASHRAE handbook, we can see that the line of wet-bulb temp = 0 has a unique trend (blue line in the picture below). It is not parallel compared to all the other wet-bulb temperature lines like wet-bulb temp = -3, -2, -1, 1, 2, 3, 4.... (red lines in the picture below). But the equations used in the function does not contain this special trend where the wet-bulb temp = 0.
image

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] ~~If any defect files are updated to a more recent version, upload new versions here or on DevSupport~~
 - [ ] ~~If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label~~
 - [ ] ~~If structural output changes, add to output rules file and add OutputChange label~~
 - [ ] ~~If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies~~

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by nrel-bot-3** [comment]
@dareumnam @lgentile it has been 28 days since this pull request was last updated.

**Comment by nrel-bot-2b** [comment]
@dareumnam @lgentile it has been 28 days since this pull request was last updated.

**Comment by dareumnam** [comment]
Investigated a few of the regression test files failed. Here are three examples of the results of the regression tests, focused on the tabular output( eplustbl.htm) and the hourly output report (eplusout.csv).

**Comment by dareumnam** [comment]
- 1ZoneEvapCooler
  - In tabular output report, 
    - Big diffs (more than 1% of percentage difference compared to the develop branch) were observed in
      - Annual Building Utility Performance Summary - End Uses
image
      - Initialization Summary - Warmup Convergence Information (of one zone, SizingPeriod)
image
      - Energy Meters - Annual and Peak Values - Water
image
      - Sensible Heat Gain Summary - Annual Building Sensible Heat Gain Components
image
    - Small diffs (less than 0.1%) were observed in 
      - Annual Building Utility Performance Summary
        - Total Site Energy and Total Source Energy
        - Total End Use - Electricity
        - Utility Use - Electricity Intensity 
      - Source Energy End Use Components Summary
        - Source Electricity End Use -Heating
      - Annual Heat Emissions Summary
        - Heat Emission - Envelope Convection, Zone Exfiltration, HVAC Relief Air, and Total
      - Energy Meters
        - Annual and Peak Values - Electricity Annual Value - Electricity:Facility, Electricity:HVAC, etc. (less than 0.02%)
        - Annual and Peak Values - Annual Value - EnergyTransfer:Facility, EnergyTransfer:Building, etc. (less than 0.05%)
  - In CSV hourly output report, (graphs show the part that showed the maximum difference throughout the year)
    - Site Outdoor Air Wetbulb Temperature
The diffs occurred when the OA Wetbulb Temperatures are below 0 C.
image
    - Baseboard Electric Power
It pops up at a few points
image
    - Mass Flow Rate of Supply Outlet Node
image
    - Zone Mean Temperature and Temperatures of several nodes (They share the same X-axis)
image

**Comment by dareumnam** [comment]
- AbsorptionChiller
  - In tabular output report, 
    - Big diffs (more than 1% of percentage difference compared to the develop branch) were observed in
      - Coil Sizing Details - Coils
image
    - Small diffs (less than 1%) were observed in
      - Annual Building Utility Performance Summary
        - Total End Use - Water [m3] (for Heat Rejection)
      - Initialization Summary - Warmup Convergence Information
        - Average Warmup Temperature Difference [deltaC]
        - Std Dev Warmup Temperature Difference [deltaC]
        - Average Warmup Load Difference [W]
        - Std Dev Warmup Load Difference [W]
      - Coil Sizing Details - Coils (in a few coils)
        - Coil Entering Air Wetbulb at Rating Conditions [C]
        - Coil Leaving Air Wetbulb at Rating Conditions [C]
      - Energy Meters
        - Annual and Peak Values - Water Annual values [m3]
 (less than 0.3%)
  - In csv hourly output report, 
    - All diffs are minor - less than 0.05%
      - Zone Air Temperatures [C] (of a few zones)
      - Plant supply side cooling demand rate [W] (of chilled/hot water loops and chilled water condenser loop)
      - Plant supply side heating demand rate [W] (of hot water loop)
      - Chiller Electric Power [W], Chiller Source Steam Rate [W], and Chiller Evaporator Cooling Rate [W] (of a chiller)

**Comment by dareumnam** [comment]
- VentilatedSlab
  - In tabular output report, 
    - Big diffs (more than 1% of percentage difference compared to the develop branch) were observed in
      - Initialization Summary - Warmup Convergence Information
image
    - Small diffs (less than 1%) were observed in
      - Initialization Summary - Warmup Convergence Information (in several zones)
        - Average Warmup Temperature Difference {deltaC}
        - Std Dev Warmup Temperature Difference {deltaC}
        - Average Warmup Load Difference {W}
      - Coil Sizing Details - Coils (in a few cooling/heating coils)
        - Coil Entering Air Wetbulb at Rating Conditions [C]
        - Coil Leaving Air Wetbulb at Rating Conditions [C]
      - Object Count Summary - Input Fields
        - Electricity Maximum Value [W] (of Electricity:Facility, ElectricityPurchased:Facility, etc.)
      - Energy Meters - Annual and Peak Values - Cooling (less than 0.05%)
        - Cooling Maximum Value [W] (of PlantLoopCoolingDemand:Facility, PlantLoopCoolingDemand:HVAC, etc.)
      - Energy Meter - Annual and Peak Values - Other (less than 0.2%)
        - Maximum Value [W] (of PlantLoopCoolingDemand:Facility, PlantLoopCoolingDemand:HVAC, etc)
  - In CSV hourly output report, (graphs show the part that showed the maximum difference throughout the year)
    - Big diffs in 
      - Mass Flow Rate of outdoor air inlet node and return air node of one zone (7.1%)
image
image
      - Zone Ventilated Slab Radiant Cooling Rate in two zones, (3.5% and 10%)
image
image

**Comment by dareumnam** [comment]
@Myoldmopar @mitchute This PR is ready for review.

File: tst/EnergyPlus/unit/DXCoils.unit.cc [hunk 1]
```diff
@@ -1355,7 +1355,7 @@ TEST_F(EnergyPlusFixture, TestMultiSpeedWasteHeat)
 
     CalcMultiSpeedDXCoilCooling(1, 1, 1, 2, 1, 1, 0);
 
-    EXPECT_NEAR(1303.4304, MSHPWasteHeat, 0.001);
+    EXPECT_NEAR(1302.7482629848564, MSHPWasteHeat, 0.001);
```
Comment (mitchute) [inline tst/EnergyPlus/unit/DXCoils.unit.cc]: In general, I think we should include (or *only include*) as many sig-figs in the expected results as we are checking for in the tolerance. So in this case, truncate the expected result to a sig-fig of 1E-3.

File: tst/EnergyPlus/unit/DXCoils.unit.cc [hunk 2]
```diff
@@ -3557,8 +3557,8 @@ TEST_F(EnergyPlusFixture, TestMultiSpeedCoilsAutoSizingOutput)
     EXPECT_EQ(0.875, DXCoil(1).MSRatedAirVolFlowRate(2) * 0.5);
     EXPECT_EQ(0.875, DXCoil(1).MSRatedAirVolFlowRate(1));
     // Design Capacity at speed 2 and speed 1
-    EXPECT_EQ(31899.037804634620, DXCoil(1).MSRatedTotCap(2));
-    EXPECT_EQ(31899.037804634620 * 0.5, DXCoil(1).MSRatedTotCap(1));
+    EXPECT_NEAR(31888, DXCoil(1).MSRatedTotCap(2), 0.01);
```
Comment (mitchute) [inline tst/EnergyPlus/unit/DXCoils.unit.cc]: Probably make the expected result ```31888.0```. Not sure if this would ever be interpreted as an ```int``` as it is currently.

File: tst/EnergyPlus/unit/DXCoils.unit.cc [hunk 3]
```diff
@@ -3810,23 +3810,23 @@ TEST_F(EnergyPlusFixture, TestMultiSpeedCoolingCoilPartialAutoSizeOutput)
...
 
     // test SHR design size when partial autosizing (capacity is hardsized)
     DXCoil( 1 ).MSRatedTotCap( 1 ) = 17500.0; // DataSizing::AutoSize;
     DXCoil( 1 ).MSRatedTotCap( 2 ) = 35000.0; // DataSizing::AutoSize;
 
     SizeDXCoil(1);
     // Design size SHR at speed 2 and speed 1
-    EXPECT_EQ(0.76488154417788068, DXCoil(1).MSRatedSHR(2));
-    EXPECT_EQ(0.76488154417788068, DXCoil(1).MSRatedSHR(1));
+    EXPECT_NEAR(0.764997, DXCoil(1).MSRatedSHR(2), 0.01);
```
Comment (mitchute) [inline tst/EnergyPlus/unit/DXCoils.unit.cc]: Match up sig-fig in the expected result to be the same as in the tolerance.

File: tst/EnergyPlus/unit/EvaporativeCoolers.unit.cc [hunk 4]
```diff
@@ -806,7 +806,7 @@ TEST_F(EvapCoolersFixture, EvaporativeCoolers_IndirectRDDEvapCoolerOperatingMode
                                                             thisEvapCooler.SecInletWetBulbTemp,
                                                             thisEvapCooler.SecInletHumRat);
     // test outlet temperature in full wet operating mode
-    EXPECT_NEAR(22.03, thisEvapCooler.OutletTemp, 0.001);
+    EXPECT_NEAR(22.035628057558771, thisEvapCooler.OutletTemp, 0.001);
```
Comment (mitchute) [inline tst/EnergyPlus/unit/EvaporativeCoolers.unit.cc]: Same here, and everywhere else.

**Comment by mitchute** [comment]
@dareumnam Thanks for the PR and fix. See the following:

- I'm not totally clear on what's happening with the first plot. We do want the functions to be smooth and continuous as you note, but I guess I don't see how that plot shows it. Maybe you could just plot this function for a few constant drybulb temps while sweeping wetbulb temp from, say, -5 to 5 C. I can probably help with that if you need, but let me know.
- A few of your plots show diffs during summer months, however, I thought this was only affecting results with wetbulb temp less than 0 C. Or maybe I'm misunderstanding. Let me know your thoughts on this.

File: tst/EnergyPlus/unit/DXCoils.unit.cc (hunk 1 above)
Comment (dareumnam) [reply tst/EnergyPlus/unit/DXCoils.unit.cc]: Thank you for letting me know. I'll fix all unit tests and update the branch again.

**Comment by dareumnam** [comment]
> @dareumnam Thanks for the PR and fix. See the following:
> 
> * I'm not totally clear on what's happening with the first plot. We do want the functions to be smooth and continuous as you note, but I guess I don't see how that plot shows it. Maybe you could just plot this function for a few constant drybulb temps while sweeping wetbulb temp from, say, -5 to 5 C. I can probably help with that if you need, but let me know.
> * A few of your plots show diffs during summer months, however, I thought this was only affecting results with wetbulb temp less than 0 C. Or maybe I'm misunderstanding. Let me know your thoughts on this.

@mitchute Thank you for your comments.
- I fixed all the unit tests with sig-figs as you commented.
- In the first plot, I was trying to show that these two equations 
image
are connected smoothly at one point, in the case of dry-bulb temp = 0 and relative humidity = 100%, where the wet-bulb temperature(t*) is around 0. (t*→ 0 + 0 and t*→ 0 - 0). 
As your comment, I plotted these functions for several constant dry-bulb temperatures: 0°C to 10°C. 
I called PsyTwbFnTdbWPb(Tdb, W, Pb) function with arguments of Dry-bulb Temp. of 0°C to 10°C (fixed value in each call), Humidity ratio that equivalent to Relative Humidity of 0% to 100% (X-axis), and Atmospheric pressure 101325 Pa (fixed value in each call). And plotted Wet-bulb temperature (Y-axis). The plotted results are in the PR description above. The result showed some gaps where the wet-bulb temperature is 0, particularly where the dry-bulb temperature is getting farther from 0.
   - I am pretty sure that the equation change itself of this PR is correct based on the ASHRAE handbook update. My best guess for these gaps is: 
According to the Psychrometric Chart in the ASHRAE handbook, we can see that the line of wet-bulb temp = 0 has a unique trend (blue line in the picture below). It is not parallel compared to all the other wet-bulb temperature lines like wet-bulb temp = -3, -2, -1, 1, 2, 3, 4.... (red lines in the picture below). But the equations used in the function does not contain this special trend where the wet-bulb temp = 0.
image
 - For the regression test results, at first, I also expected that it might only affect results with wet-bulb temp less than 0°C. But I realized that the function was changed not only the equation for wet-bulb temp < 0 but also the equation for wet-bulb temp ≥ 0. I guess this might bring different results with wet-bulb temp more than 0. Even though the wet-bulb diffs were small, they might affect other results after the cumulation of each time step, which seemed big. I could investigate this in detail with one or a few example regression test files if needed. Please let me know.

**Comment by mitchute** [comment]
@dareumnam thanks for the updates.

I went ahead and plotted Twb vs. Tdb for a couple of different humidity ratios and compared the results to [CoolProp](http://www.coolprop.org/fluid_properties/HumidAir.html#). See the plots below.

image

image

The results compare well, except near the Twb > 0 transition point where the functions become discontinuous which causes steep derivatives.

@Myoldmopar @mjwitte I'm comfortable that this is working as expected. However, do you think it would be beneficial to add some smoothing between these two functions, say, from Twb = -1 to 1 C? Or some other range? Or do we just ignore it and move on?

**Comment by mitchute** [comment]
@dareumnam could you pull develop into this branch one more time?

**Comment by mitchute** [comment]
@Myoldmopar it looks like CI is hung up on this. All of the CI machines have made about three passes on the most recent commit.

**Comment by Myoldmopar** [comment]
Eww.  I wonder what happened.  I'll take a look and also do a final review on this.  Thanks!

**Comment by mitchute** [comment]
CI is complaining a lot here. I want to say that the diffs are justified simply due to the straightforwardness of the changes, however, I need to take a closer look. For example in the ```5ZoneBranchSupplyPumps.idf``` file, why are there summer design day diffs when these calcs should only affect cases when Tdb is less than 0 C. See below.

Var_16

(Apologies. The horizontal axis is simply the index of the csv file. The summer design day begins at index 26)

OK, the diffs are _tiny_, but still, why are they there to begin with? I think once I plot a few more cases I'll feel more comfortable with this.

And there's still the issue of whether we need to apply some smoothing. See this https://github.com/NREL/EnergyPlus/pull/7548#issuecomment-577868676 

@Myoldmopar @mjwitte tag on here when you can.

**Comment by Myoldmopar** [comment]
I feel like this has been discussed and fixed up long enough.  I'm not too worried about that smoothing, it generally looks to be a good match with cool prop, and the diffs are all over the place, but seems to be OK.  The unit test changes to make this pass look exactly like I expected.  I am going to pull in develop so that Linux 7.5 works on this this morning and we'll go from there.

**Comment by mitchute** [comment]
There was a conflict so I pulled develop in again. Assuming no unit test failures, this can merge. The other errors are going to be expected given what this is changing.

**Comment by Myoldmopar** [comment]
Thanks for pulling develop in again @mitchute.  I'm going to time the merge of this one carefully since it causes so much CI mess.  I am hoping to merge a few clean PRs this morning, then pull develop into a few branches and let CI run this afternoon.  Then if this is ready, it will merge late afternoon.  Does that sound OK?

**Comment by mitchute** [comment]
@Myoldmopar fine by me. These results are as good as it's going to get, so we might as well get it off queue when we can.

**Comment by Myoldmopar** [comment]
OK, merging this in.  It's a very large set of diffs, but they've been dug through multiple times.  The functional change is so small, it's amazing it's taken this much effort.  Just a couple lines changed, followed by a couple handful of unit test changes -- mostly where precision was too tight anyway --, and that's it.  This is good.
